### PR TITLE
feat(git): support `repo//subdir` URL syntax for sub-directory deps (#509)

### LINKED ISSUE #509: feature: Add a module under the git repository's sub-directory as a dependency
## Feature Request

Add a module under the git repository's sub-directory as a dependency.

### PR DESCRIPTION
**Pull request by Peefy**
## Summary

Lets users depend on a single KCL package that lives inside a larger git repository, without having to clone the whole tree.

## How

`kcl mod add` and `kcl run` now understand the go-getter `//` sub-directory selector. A URL of the form

```
https://github.com/org/monorepo.git//packages/kcl/foo
```

correctly extracts `packages/kcl/foo` as the sub-directory and routes the clone through go-getter, which only materialises that subtree.

Supported variants:

- plain: `https://host/repo.git//path/to/pkg`
- with scheme prefix: `git::https://host/repo.git//path`
- with ref and subdir: `git::https://host/repo.git?ref=v1.0//path`
- query on subdir: `https://host/repo.git//path?ref=v1` (ref is preserved on base)
- fragment: `https://host/repo.git//path#frag`

## Files

- `pkg/git/getter.go`: new `SplitSubdir(repoURL) (base, sub, err)` parser + `isLikelyScheme` helper. `Clone()` routes the URL through `SplitSubdir` before handing it to go-getter.
- `pkg/git/getter_test.go` *(new)*: table-driven tests covering scheme preservation, query/fragment forwarding, trailing-slash edge cases, and the `https://` / `ssh://` scheme-delimiter collision.

## Test plan

- [x] `go test ./pkg/git/...` green locally (12 sub-cases in `TestSplitSubdir`).
- [ ] CI on the branch (no e2e impact).

Closes #509.

### CODE REVIEW

**Comment by Peefy** [comment]
## CI fix for `SplitSubdir`

The `Unit tests with coverage` job is red, and it isn't just one test — **8 top-level tests fail**, all git-related, all from a single root cause in `SplitSubdir`:

```
TestRun/TestRunWithGitPackage
TestAddGitNewLocalStorage
TestUpdateGitNewLocalStorage
TestCloneWithOptions/LocalBareCloneAsNonBareWithCommit
TestKpmIssue226   (add/update with git commit)
TestKpmIssue550   (SupportNewStorage, SupportNewStorageAndMVS)
TestKpmIssue587   (test download with git dep)
TestWithGlobalLock/TestGitDownloader
```

### Root cause

`SplitSubdir` searches the whole URL with `strings.LastIndex(body, "//")`. That matches `//` in two places where it is *not* a sub-directory selector:

1. **`file:///abs/path`** — the leading `///` (empty authority followed by an absolute path) is matched, so the absolute path is sliced off as a "subdir".
2. **`?ref=<hash>//<path>`** — a `//` that lives inside a *query value* is treated as a subdir selector.

Here is the smoking gun from the failing job:

```
error downloading 'file:?ref=ade147b%2F%2Ftmp%2F.../flask-demo-kcl-manifests'
/usr/bin/git exited with 128
ssh: Could not resolve hostname file: Temporary failure in name resolution
```

The base collapsed to a bare **`file:`**. Tracing case 1 on the local bare-repo cache URL `file:///tmp/.../flask-demo-kcl-manifests?ref=ade147b`:

- `LastIndex("//")` lands inside the `file:///` prefix
- the `body[idx-1] == ':'` guard does not trip, because the preceding byte is `/`, not `:`
- `base = strings.TrimRight("file:/", "/")` → `"file:"`
- the whole path becomes the subdir, and `?ref=...` is re-appended to that stub base

git then reads `file:` as an scp-style ssh target and tries to resolve a host literally named `file`. That is why every git-backed test dies the same way — and why the Windows job happens to pass, since it takes a different path.

### Fix

Three changes in `SplitSubdir`:

| Change | Why |
| --- | --- |
| Slice the `?query` / `#fragment` suffix off into `tail` **before** searching, re-attach it on return | a `//` inside a query or fragment value can never be a subdir selector |
| Anchor the search to the URL **path** — the first `/` after `://<authority>` | the `://` scheme delimiter and the leading `///` of `file:///abs/path` are never matched |
| Drop the magic `idx < 16` cap on scheme detection and rely on `isLikelyScheme` alone | any valid forced scheme (`git::`, `hg::`, …) is accepted, while a path that merely contains `::` is still rejected |

### One existing test case retargeted

```go
- in: "git::https://github.com/org/repo.git?ref=v1.0//pkg/sub"
+ in: "git::https://github.com/org/repo.git//pkg/sub?ref=v1.0"
```

`?ref=...//subdir` is genuinely ambiguous — nothing can tell whether that `//` is a subdir selector or part of the query value. Subdir-then-ref is the canonical go-getter form. A comment in the test documents the limitation explicitly.

### Verification

- Extracted `SplitSubdir` + `isLikelyScheme` into a standalone module and ran this PR's own test file: **30/30 subtests pass**
- `gofmt -l pkg/git/` clean after applying — the patch also aligns the `cases` struct fields in `getter_test.go`, folded in here since that file is new in this PR
- No new imports

7 regression cases added, including the exact URL from the CI failure, plus `ssh://git@host/...//pkg` (userinfo in the authority), `git::https://...` with no path, and `//` inside a fragment.

<details>
<summary>Patch — <code>git apply</code> from the repo root</summary>

```diff
diff --git a/pkg/git/getter.go b/pkg/git/getter.go
index d304ce2..b70e90d 100644
--- a/pkg/git/getter.go
+++ b/pkg/git/getter.go
@@ -30,48 +30,74 @@ func ForceProtocol(url, protocol string) string {
 //
 //	SplitSubdir("https://github.com/org/repo.git//pkg/sub")
 //	  -> "https://github.com/org/repo.git", "pkg/sub", nil
-//	SplitSubdir("git::https://github.com/org/repo.git?ref=v1//pkg/sub")
+//	SplitSubdir("git::https://github.com/org/repo.git//pkg/sub?ref=v1")
 //	  -> "git::https://github.com/org/repo.git?ref=v1", "pkg/sub", nil
 //	SplitSubdir("https://github.com/org/repo.git")
 //	  -> "https://github.com/org/repo.git", "", nil
-//	SplitSubdir("https://github.com/org/repo.git//")
-//	  -> "https://github.com/org/repo.git//", "", nil  (no content after //)
+//	SplitSubdir("file:///abs/path?ref=abc//subdir")
+//	  -> "file:///abs/path?ref=abc//subdir", "", nil
 //
-// The split is intentionally conservative: only the LAST `//` that is
-// not part of a scheme delimiter (i.e. not preceded by ":") and that
-// is followed by a non-empty path component is treated as a
-// sub-directory separator.
+// The split is intentionally conservative:
+//   - The "?query" / "#fragment" suffix is sliced off before searching
+//     for the subdir separator, so a "//" that lives inside a query or
+//     fragment value (e.g. `?ref=<hash>//subdir`) is never mistaken for
+//     a subdir selector.
+//   - The search is anchored to the URL path (after the `://<authority>`
+//     delimiter), so the leading "///" of `file:///abs/path` and the
+//     "://" scheme delimiter are never matched.
 func SplitSubdir(repoURL string) (string, string, error) {
 	if repoURL == "" {
 		return "", "", nil
 	}
 
 	// Schemes like "git::" can confuse a naive split, so split off the
-	// leading scheme if present and re-attach after parsing.
+	// leading scheme if present and re-attach after parsing. We rely
+	// on isLikelyScheme (letters/digits/_/-) to validate the prefix
+	// rather than a length cap, so any go-getter forced scheme is
+	// recognised and a path that merely contains "::" is left alone.
 	schemePrefix := ""
 	body := repoURL
-	if idx := strings.Index(repoURL, "::"); idx >= 0 && idx < 16 {
-		candidate := repoURL[:idx]
-		if isLikelyScheme(candidate) {
-			schemePrefix = repoURL[:idx+2]
-			body = repoURL[idx+2:]
-		}
+	if idx := strings.Index(repoURL, "::"); idx >= 0 && isLikelyScheme(repoURL[:idx]) {
+		schemePrefix = repoURL[:idx+2]
+		body = repoURL[idx+2:]
 	}
 
-	// Find the LAST "//" occurrence. Earlier occurrences (e.g. "https://")
-	// must be skipped.
-	idx := strings.LastIndex(body, "//")
-	if idx < 0 {
-		return repoURL, "", nil
+	// Slice off any "?query" or "#fragment" before searching for the
+	// subdir separator. A "//" inside the query/fragment value is not
+	// a subdir selector — this is the bug that broke URLs of the form
+	// `file:///<cache>/repo?ref=<hash>//subdir` in TestRunWithGitPackage.
+	suffixStart := len(body)
+	for i := 0; i < len(body); i++ {
+		if body[i] == '?' || body[i] == '#' {
+			suffixStart = i
+			break
+		}
+	}
+	head := body[:suffixStart]
+	tail := body[suffixStart:]
+
+	// Restrict the "//" search to the URL path so the scheme delimiter
+	// (`://`) and the leading `///` of `file:///abs/path` (where the
+	// authority is empty) are never matched. Without this, the rightmost
+	// "//" in `file:///tmp/foo` lands inside the authority/path boundary
+	// and splits the absolute path off as a "subdir".
+	searchFrom := 0
+	if sIdx := strings.Index(head, "://"); sIdx >= 0 {
+		rest := head[sIdx+3:]
+		if pIdx := strings.Index(rest, "/"); pIdx >= 0 {
+			searchFrom = sIdx + 3 + pIdx
+		} else {
+			searchFrom = len(head)
+		}
 	}
 
-	// We only treat this as a sub-dir separator if it is NOT part of a
-	// scheme delimiter (i.e. not preceded by ":").
-	if idx > 0 && body[idx-1] == ':' {
+	idx := strings.LastIndex(head[searchFrom:], "//")
+	if idx < 0 {
 		return repoURL, "", nil
 	}
+	idx += searchFrom
 
-	rawSub := body[idx+2:]
+	rawSub := head[idx+2:]
 	if rawSub == "" {
 		// Trailing "//" with nothing after — there is no sub-directory to
 		// extract. Return the URL unchanged so the caller sees the
@@ -79,8 +105,14 @@ func SplitSubdir(repoURL string) (string, string, error) {
 		return repoURL, "", nil
 	}
 
-	base := strings.TrimRight(body[:idx], "/")
-	sub := strings.TrimLeft(rawSub, "/")
+	base := strings.TrimRight(head[:idx], "/")
+	sub := rawSub
+
+	// Belt-and-braces: a "//" preceded by ":" would be a scheme
+	// delimiter we somehow missed (e.g. URL with no path).
+	if idx > 0 && head[idx-1] == ':' {
+		return repoURL, "", nil
+	}
 
 	// The "sub" portion may itself contain a query string or fragment
 	// because users sometimes write "...//subdir?ref=v1". Split it off
@@ -99,11 +131,11 @@ func SplitSubdir(repoURL string) (string, string, error) {
 		// Subdir portion was purely query/fragment — treat as no subdir.
 		return repoURL, "", nil
 	}
-	return schemePrefix + base, sub, nil
+	return schemePrefix + base + tail, sub, nil
 }
 
 // isLikelyScheme reports whether `s` looks like a go-getter scheme name
-// (letters, digits, underscores).
+// (letters, digits, underscores, hyphens).
 func isLikelyScheme(s string) bool {
 	if s == "" {
 		return false
diff --git a/pkg/git/getter_test.go b/pkg/git/getter_test.go
index 531bed1..6c8f35c 100644
--- a/pkg/git/getter_test.go
+++ b/pkg/git/getter_test.go
@@ -8,10 +8,10 @@ import (
 
 func TestSplitSubdir(t *testing.T) {
 	cases := []struct {
-		name      string
-		in        string
-		wantBase  string
-		wantSub   string
+		name     string
+		in       string
+		wantBase string
+		wantSub  string
 	}{
 		{
 			name:     "no subdir",
@@ -37,9 +37,13 @@ func TestSplitSubdir(t *testing.T) {
 			wantBase: "git::https://github.com/org/repo.git",
 			wantSub:  "pkg",
 		},
+		// Subdir-then-ref is the canonical go-getter form.
+		// `?ref=...//subdir` is ambiguous and is not supported:
+		// SplitSubdir cannot tell whether the `//` is a subdir
+		// selector or part of the query value.
 		{
-			name:     "scheme with ref and subdir",
-			in:       "git::https://github.com/org/repo.git?ref=v1.0//pkg/sub",
+			name:     "scheme with subdir then ref",
+			in:       "git::https://github.com/org/repo.git//pkg/sub?ref=v1.0",
 			wantBase: "git::https://github.com/org/repo.git?ref=v1.0",
 			wantSub:  "pkg/sub",
 		},
@@ -85,6 +89,53 @@ func TestSplitSubdir(t *testing.T) {
 			wantBase: "https://example.com/repo",
 			wantSub:  "a/b/c/d",
 		},
+		// Regression: file:///abs/path was previously split into
+		// `file:` + `//abs/path` because the leading `///` was
+		// matched by LastIndex("//").
+		{
+			name:     "file scheme with absolute path",
+			in:       "file:///tmp/foo/bar",
+			wantBase: "file:///tmp/foo/bar",
+			wantSub:  "",
+		},
+		{
+			name:     "file scheme with ref and no subdir",
+			in:       "file:///tmp/foo/bar?ref=ade147b",
+			wantBase: "file:///tmp/foo/bar?ref=ade147b",
+			wantSub:  "",
+		},
+		// Regression: this is the exact URL shape that broke
+		// TestRunWithGitPackage in the original PR. The `?ref=...`
+		// belongs to the base URL; the trailing `//tmp/...` is
+		// inside the query value and must NOT be treated as a subdir.
+		{
+			name:     "file scheme with ref whose value contains //",
+			in:       "file:///tmp/cache/flask-demo-kcl-manifests?ref=ade147b//tmp/cache/flask-demo-kcl-manifests",
+			wantBase: "file:///tmp/cache/flask-demo-kcl-manifests?ref=ade147b//tmp/cache/flask-demo-kcl-manifests",
+			wantSub:  "",
+		},
+		// ssh URLs have user@ in the authority; the search must
+		// still anchor past the authority delimiter.
+		{
+			name:     "ssh scheme with userinfo and subdir",
+			in:       "ssh://git@github.com/org/repo.git//pkg",
+			wantBase: "ssh://git@github.com/org/repo.git",
+			wantSub:  "pkg",
+		},
+		// Forced scheme with an empty path is preserved verbatim.
+		{
+			name:     "scheme prefix preserved when body has no path",
+			in:       "git::https://github.com/org/repo.git",
+			wantBase: "git::https://github.com/org/repo.git",
+			wantSub:  "",
+		},
+		// `//` inside a fragment value is not a subdir selector.
+		{
+			name:     "fragment value contains //",
+			in:       "https://host/path#frag//with//slashes",
+			wantBase: "https://host/path#frag//with//slashes",
+			wantSub:  "",
+		},
 	}
 	for _, c := range cases {
 		t.Run(c.name, func(t *testing.T) {
@@ -126,6 +177,11 @@ func TestIsLikelyScheme(t *testing.T) {
 		{"", false},
 		{"git repo", false},
 		{"git.proto", false},
+		// Long but valid scheme names must now be accepted (the
+		// previous 16-char cap is gone).
+		{"verylongschemename", true},
+		// A path with "::" must not be mis-classified as a scheme.
+		{"path/with/colons", false},
 	}
 	for _, c := range cases {
 		t.Run(c.in, func(t *testing.T) {
@@ -134,4 +190,4 @@ func TestIsLikelyScheme(t *testing.T) {
 			}
 		})
 	}
-}
\ No newline at end of file
+}
```

</details>
